### DISCUSSION TITLE
[0.75] Autolinking performance benefits shipping in React Native 0.75 & Expo SDK 51

### DISCUSSION THREAD

**Discussion by Kudo** [react-native-community/discussions-and-proposals#814] [Announcements]
**TL;DR: The upcoming stable version of React Native, 0.75, will come with significant improvements to the auto-linking algorithm.**

Hi all,

We (Expo) would like to share some performance wins that are shipping with 0.75 which we worked on together with folks from the React Native team at Meta.

As part of the ongoing effort to [define React Native framework (RFC759)](https://github.com/react-native-community/discussions-and-proposals/blob/main/proposals/0759-react-native-frameworks.md) and enhance the core system to better support all React Native frameworks, version 0.75 introduces a complete rewrite of the autolinking implementation, moving some of its functionalities into the core system. 

Autolinking was first [introduced in React Native 0.60](https://reactnative.dev/blog/2019/07/03/version-60), and has helped significantly developers to just include new libraries in their project with a simple `install` invocation. Since 0.60 the auto-linking algorithm has been living inside the Community CLI and has been improved significantly thanks to various community contributions.

However, during the work for RFC759, we noticed the auto-linking responsibilities could be better split. So with 0.75 the autolinking logic is split as follows:

- React Native core is now responsible for generating the code needed to link third-party modules into both the CocoaPods and Gradle build systems.
- The discovery of third-party modules is now a responsibility of frameworks, with the core accepting a JSON config to define linked modules.

By default, your projects will use the `@react-native-community/cli` to discover modules, specifically with the `config` command. So these projects will keep on working as usual, and your modules will be correctly linked.

Splitting the autolinking logic allowed us to unlock performance gains, which are particularly significant for bigger projects with a lot of dependencies.

Specifically if you’re using Expo, the autolinking discovery logic has been rewritten to be ready for official adoption in the near future. 

According to our benchmarks (see below), the new autolinking implementation runs between 1.35 and 1.64 times as fast on iOS, and between 2.88 and 10.24 times as fast on Android compared to the previous autolinking in React Native 0.74.

If you are using the latest Expo SDK in your projects, you can enable this new autolinking feature by setting the environment variable `EXPO_UNSTABLE_CORE_AUTOLINKING=1`. 

Please give it a try and experience the improvements firsthand.

# Benchmarks

### On the [bare-expo app](https://github.com/expo/expo/tree/main/apps/bare-expo)

Android is 2.88 times faster and iOS is 1.64 times faster.

```
$ hyperfine 'npx @react-native-community/cli config --platform android' 'npx expo-modules-autolinking react-native-config --platform android --json'

Benchmark 1: npx @react-native-community/cli config --platform android
  Time (mean ± σ):      2.150 s ±  0.175 s    [User: 1.532 s, System: 0.653 s]
  Range (min … max):    2.003 s …  2.598 s    10 runs

Benchmark 2: npx expo-modules-autolinking react-native-config --platform android --json
  Time (mean ± σ):     747.3 ms ±  15.0 ms    [User: 854.3 ms, System: 602.3 ms]
  Range (min … max):   730.7 ms … 771.9 ms    10 runs

Summary
  npx expo-modules-autolinking react-native-config --platform android --json ran
    2.88 ± 0.24 times faster than npx @react-native-community/cli config --platform android
```

```
$ hyperfine 'npx @react-native-community/cli config --platform ios' 'npx expo-modules-autolinking react-native-config --platform ios --json'

Benchmark 1: npx @react-native-community/cli config --platform ios
  Time (mean ± σ):     844.3 ms ±  89.9 ms    [User: 774.4 ms, System: 154.2 ms]
  Range (min … max):   793.4 ms … 1097.2 ms    10 runs

  Warning: The first benchmarking run for this command was significantly slower than the rest (1.097 s). This could be caused by (filesystem) caches that were not filled until after the first run. You should consider using the '--warmup' option to fill those caches before the actual benchmark. Alternatively, use the '--prepare' option to clear the caches before each timing run.

Benchmark 2: npx expo-modules-autolinking react-native-config --platform ios --json
  Time (mean ± σ):     513.9 ms ±   8.2 ms    [User: 520.2 ms, System: 99.9 ms]
  Range (min … max):   504.9 ms … 530.7 ms    10 runs

Summary
  npx expo-modules-autolinking react-native-config --platform ios --json ran
    1.64 ± 0.18 times faster than npx @react-native-community/cli config --platform ios
```

### On the [react-conf app](https://github.com/expo/react-conf-app)

Android is 10.24 times faster and iOS is 1.35 times faster.

```
$ hyperfine 'npx @react-native-community/cli config --platform android' 'npx expo-modules-autolinking react-native-config --platform android --json'

Benchmark 1: npx @react-native-community/cli config --platform android
  Time (mean ± σ):      3.369 s ±  0.054 s    [User: 3.256 s, System: 0.151 s]
  Range (min … max):    3.318 s …  3.475 s    10 runs

Benchmark 2: npx expo-modules-autolinking react-native-config --platform android --json
  Time (mean ± σ):     328.9 ms ±  10.9 ms    [User: 407.9 ms, System: 113.6 ms]
  Range (min … max):   319.8 ms … 358.0 ms    10 runs

Summary
  npx expo-modules-autolinking react-native-config --platform android --json ran
   10.24 ± 0.38 times faster than npx @react-native-community/cli config --platform android
```

```
$ hyperfine 'npx @react-native-community/cli config --platform ios' 'npx expo-modules-autolinking react-native-config --platform ios --json'

Benchmark 1: npx @react-native-community/cli config --platform ios
  Time (mean ± σ):     335.5 ms ±   2.4 ms    [User: 352.2 ms, System: 77.9 ms]
  Range (min … max):   331.3 ms … 339.5 ms    10 runs

Benchmark 2: npx expo-modules-autolinking react-native-config --platform ios --json
  Time (mean ± σ):     248.4 ms ±   7.1 ms    [User: 263.4 ms, System: 53.8 ms]
  Range (min … max):   243.6 ms … 268.8 ms    11 runs

Summary
  npx expo-modules-autolinking react-native-config --platform ios --json ran
    1.35 ± 0.04 times faster than npx @react-native-community/cli config --platform ios
```

# Acknowledgement
We want to acknowledge the folks on the React Native team at Meta for working on this autolinking refactoring which enabled those performance savings, as well as the Community CLI maintainers for supporting this refactoring.

**Comment by mrousavy** [comment]
nice work! 👏

**Comment by adsalihac** [comment]
😍
